mailbox/noise.go

```diff
@@ -321,27 +345,391 @@ func newHandshakeState(initiator bool, prologue []byte,
 	// TODO(roasbeef): if did mixHash here w/ the password, then the same
 	// as using it as a PSK?
 
-	return h
+	// Call mixHash once for every pub key listed in the pre-messages from
+	// the handshake pattern.
+	for _, m := range pattern.PreMessages {
+		if m.Initiator == h.initiator {
+			h.mixHash(localPub.PubKey().SerializeCompressed())
+		} else {
+			if remoteStatic == nil {
+				return handshakeState{},
+					errors.New("a remote static key is " +
+						"expected for the given " +
+						"handshake pattern")
+			}
+			h.mixHash(remoteStatic.SerializeCompressed())
+		}
+	}
+
+	return h, nil
 }
 
-// EphemeralGenerator is a functional option that allows callers to substitute
-// a custom function for use when generating ephemeral keys for ActOne or
-// ActTwo.  The function closure return by this function can be passed into
-// NewBrontideMachine as a function option parameter.
-func EphemeralGenerator(gen func() (*btcec.PrivateKey, error)) func(*Machine) {
-	return func(m *Machine) {
-		m.ephemeralGen = gen
+// writeMsgPattern uses the given MessagePattern to construct the data stream
+// to write to the passed writer. First the handshake version is written, then
+// the MessagePattern Tokens are processed and then any payload is encrypted
+// in a way defined by the handshake version.
+func (h *handshakeState) writeMsgPattern(w io.Writer, mp MessagePattern) error {
+	buff := new(bytes.Buffer)
+
+	// Write handshake byte to the buffer.
+	if _, err := buff.Write([]byte{h.version}); err != nil {
+		return err
+	}
+
+	// Write Tokens.
+	if err := h.writeTokens(mp.Tokens, buff); err != nil {
+		return err
+	}
+
+	// Write payload data.
+	switch h.version {
+	case HandshakeVersion:
+		var payload []byte
+		switch mp.ActNum {
+		case act1, act3:
+		case act2:
+			// If we have an auth payload, then we'll write
+			// out 2 bytes that denotes the true length of
+			// the payload, followed by the payload itself.
+			var payloadWriter bytes.Buffer
+			payload = make([]byte, ActTwoPayloadSize)
+
+			if h.payloadToSend != nil {
+				var length [lengthHeaderSize]byte
+				payLoadLen := len(h.payloadToSend)
+				binary.BigEndian.PutUint16(
+					length[:], uint16(payLoadLen),
+				)
+
+				_, err := payloadWriter.Write(length[:])
+				if err != nil {
+					return err
+				}
+
+				_, err = payloadWriter.Write(h.payloadToSend)
+				if err != nil {
+					return err
+				}
+			}
+
+			copy(payload[:], payloadWriter.Bytes())
+
+		default:
+			return fmt.Errorf("unknown act number: %d", mp.ActNum)
+		}
+
+		_, err := buff.Write(h.EncryptAndHash(payload))
```

Comment:
Getting through the PR still, but marking here to ensure that the other party is actually able to read this information as is. As right now, I don't see the existing encrypted length prefix that we use (that includes a MAC) as normal.